src/rosdiscover/interpreter/context.py

```diff
@@ -29,8 +29,8 @@ class NodeContext:
     _placeholder: bool = attr.ib(default=False, repr=False)
     _uses: Set[Tuple[str, str]] = attr.ib(factory=set, repr=False)
     _provides: Set[Tuple[str, str]] = attr.ib(factory=set, repr=False)
-    _subs: Set[Tuple[str, str]] = attr.ib(factory=set, repr=False)
-    _pubs: Set[Tuple[str, str]] = attr.ib(factory=set, repr=False)
+    _subs: Set[Tuple[str, str, bool]] = attr.ib(factory=set, repr=False)
```

Comment:
This is getting very stinky and dangerous. (`Tuple[str, str]` already wasn't great.) Every thing that interacts with `_subs` now needs to be changed and it's now easy to confuse the order of the items in the tuple and create sneaky new errors. The proper fix here is to introduce a new `Topic` data class with `name`, `type_`, and `implicit` parameters.

Reply:
Yeah. I've undone all this in this branch and I think you can now merge.